### ISSUE TITLE
[TypeSpec Validation] Show per-rule status in GitHub Checks UI

### ISSUE THREAD

**Issue by mikeharder** [Azure/azure-rest-api-specs#24999]
To make it easier for spec authors to interpret the results, TypeSpecValidation should output the status of each rule directly to the GitHub Checks UI, similar to the previous tool:

![image](https://github.com/Azure/azure-rest-api-specs/assets/9459391/8663d3e8-85b4-4e21-a249-5e08bfb5fde0)
https://github.com/Azure/azure-rest-api-specs/pull/24937/checks?check_run_id=15378216283

I believe this is implemented via the `appendMsg:` statements in the log:

```
appendMsg:{"type":"Raw","level":"Error","message":"InConsistentSwagger","time":"2023-07-26T23:05:04.813Z","extra":{"
```
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2950943&view=logs&j=33174e5f-e491-5dfc-305d-c09df824202b&t=0f377f82-3711-5097-078f-1ce87e974d2b&l=930

```
appendMsg:[{"type":"Result","code":"@azure-tools/typespec-azure-core/documentation-required","level":"Warning","message":"The ModelProperty named 'filter'
```
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2950943&view=logs&j=33174e5f-e491-5dfc-305d-c09df824202b&t=0f377f82-3711-5097-078f-1ce87e974d2b&l=6080

**Comment by konrad-jamrozik** [comment]
As discussed with @mikeharder, we also want to have "getting help" link, as captured by this work item:

- https://github.com/Azure/azure-sdk-tools/issues/6286

Which fits in the bigger picture captured here:

- https://github.com/Azure/azure-sdk-tools/issues/6621

@ckairen for the current implementation on how the table is composed, the call site for that is `renderUnifiedPipelineCheck.ts` / `buildCheckResult` [calling into `buildCheckText`](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=private/openapi-kebab/src/bots/pipeline/pipelineEventListener/renderUnifiedPipelineCheck.ts&version=GCda8fe4693ff3e40b4366284160d376f068719154&line=124&lineStartColumn=19&lineEndColumn=33&_a=contents)

which then calls:

`const generateView = compileHandlebarsTemplate<HandlebarCheckGenData>("checker.handlebars", { isDetail: true });` from `checkerBuilder.ts` which uses the [`checker.handlebars`](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=private/openapi-kebab/src/bots/pipeline/templates/checker.handlebars&version=GC788d23e724120a0477cc5177d4aafeed40127700&_a=contents) template.

**Comment by mikeharder** [comment]
@konrad-jamrozik: Are you saying that `appendMsg` in the logs is something specific to `pipeline-bot` and not a general-purpose feature to set UI in GitHub status checks?

If so, do you know the general-purpose low-level feature that can be used to set UI in GitHub status checks?

**Comment by mikeharder** [comment]
> If so, do you know the general-purpose low-level feature that can be used to set UI in GitHub status checks?

Maybe this API?

https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run--parameters

And I think we could create a GitHub Action to call the API.  Could the GitHub action access the raw log of the TSV check, convert it to a nice markdown UI, and call the checks API to set the nice UI?

**Comment by konrad-jamrozik** [comment]
@mikeharder yes, [here](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=public/rest-api-specs-scripts/src/unifiedPipelineHelper.ts&version=GC788d23e724120a0477cc5177d4aafeed40127700&line=130&lineStartColumn=11&lineEndColumn=20&_a=contents) and [here](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=private/azure-swagger-validation/azureSwaggerValidation/src/unifiedPipelineHelper.ts&version=GC5ed19c8161d4cce5bbb6d511da224d5ce821e815&line=229&lineStartColumn=11&lineEndColumn=20&_a=contents) are the `appendMsg` sources.

For setting check status, conclusion and the various part of description, see [this code I wrote recently](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=private/openapi-kebab/src/bots/pipeline/pipelineEventListener/renderAutomatedMergingRequirementsMetCheck.ts&version=GC33f9b2a20a09e61c00930b0383f6253e9a81876a&line=130&lineEnd=134&lineStartColumn=1&lineEndColumn=109&_a=contents). 

GitHub APIs used in the code above:

- https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run
- https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#update-a-check-run

but this is only for writing. Unified-pipeline never reads the check values, it only writes them.

When it comes to reading the check values the story is a bit more complicated. There are status checks ([which have two subtypes: statuses and... checks](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/about-status-checks#types-of-status-checks-on-github)), check runs and check suites. [Doc](https://docs.github.com/en/apps/creating-github-apps/writing-code-for-a-github-app/building-ci-checks-with-a-github-app#about-checks). I had a call with @benbp where explained to me some of the complexities. I can share the recording. I will be also diving into the problem head-first in the next 2 weeks to address these two work items:

- https://github.com/Azure/azure-sdk-tools/issues/6613
- https://github.com/Azure/azure-sdk-tools/issues/6617

**Comment by konrad-jamrozik** [comment]
@mikeharder re:

> And I think we could create a GitHub Action to call the API. Could the GitHub action access the raw log of the TSV check, convert it to a nice markdown UI, and call the checks API to set the nice UI?

I believe so. The interesting question here is about the nice markdown UI. If you want to have a table looking exactly like in our unified pipeline-based checks, then the code for that is as I explained in [my previous comment](https://github.com/Azure/azure-rest-api-specs/issues/24999#issuecomment-1670720655). That code is tightly coupled with unified pipeline.

My recommendation would be to write new, independent code to do the table formatting, even if the table will end up looking a little bit differently. Doesn't matter IMO. We can use the existing unified pipeline code as an inspiration, but use this as an opportunity for a rewrite.

**Comment by konrad-jamrozik** [comment]
Here are some links I got from @benbp:

> [azure-sdk-actions/types.go at main · Azure/azure-sdk-actions (github.com)](https://github.com/Azure/azure-sdk-actions/blob/main/types.go)
[azure-sdk-actions/testpayloads at main · Azure/azure-sdk-actions (github.com)](https://github.com/Azure/azure-sdk-actions/tree/main/testpayloads)
where I parse check suite url from PR json data: [azure-sdk-actions/types.go at main · Azure/azure-sdk-actions (github.com)](https://github.com/Azure/azure-sdk-actions/blob/main/types.go#L143)
iterate check suites, get the IDs, call check runs. Each check suite json should have a url for its check runs that you can use

I am now working on getting the required checks info from GitHub in a way where I can also pull the `TypeSpec Validation` status into the `Automated merging requirements met` check, not only unified pipleline-based checks.

**Comment by mikeharder** [comment]
> @mikeharder Mike Harder FTE re:
> 
> > And I think we could create a GitHub Action to call the API. Could the GitHub action access the raw log of the TSV check, convert it to a nice markdown UI, and call the checks API to set the nice UI?
> 
> I believe so. The interesting question here is about the nice markdown UI. If you want to have a table looking exactly like in our unified pipeline-based checks, then the code for that is as I explained in [my previous comment](https://github.com/Azure/azure-rest-api-specs/issues/24999#issuecomment-1670720655). That code is tightly coupled with the unified pipeline infrastructure which includes Cosmos DB instance, workflow that includes processing of events through event hub, a lot of custom types (like the [Result](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/openapi-alps?path=private/openapi-kebab/src/bots/pipeline/msgInterfaceUtils.ts&version=GC0ab4a859066e3aad63869f27429842591f80aae8&line=17&lineStartColumn=18&lineEndColumn=24&_a=contents)) and bunch of other things.
> 
> My recommendation would be to write new, independent code to do the row data plumbing and table formatting, even if the table will end up looking a little bit differently. Doesn't matter IMO. We can use the existing unified pipeline code as an inspiration, but do not try to reuse it. At most: copy-paste-adapt-refactor.
> 
> It would just be cool if we try to write as much logic as possible in C# or TypeScript, not PowerShell. Keep the plumbing in PowerShell inside the GitHub action to absolute minimum we can get away with. I am sure we will end up with significant amount of code in the end, especially given we plan on migrating more things over from unified pipeline.

I agree with re-writing the component to convert from raw pipeline logs to nicely-formatted information directly in the PR.  The format doesn't need to exactly match the previous checks, just be reasonable.  Also agree with using TypeScript/JavaScript, since this seems like the de-facto language for GitHub Actions.

I'm still not sure if/how a GitHub action can read the raw pipeline logs, or if we'd need a bot to flow data from pipeline logs to GitHub Action (using an intermediate like EventHubs).

**Comment by konrad-jamrozik** [comment]
@mikeharder here is an example of multiple lines coming from ADO log to GitHub check view:

https://github.com/Azure/azure-rest-api-specs/pull/24548/checks?check_run_id=15419664559

![image](https://github.com/Azure/azure-rest-api-specs/assets/4429827/bb41f74c-8af1-4dbc-8f04-e12f00df7c38)

**Comment by mikeharder** [comment]
> @mikeharder Mike Harder FTE here is an example of multiple lines coming from ADO log to GitHub check view:
> 
> https://github.com/Azure/azure-rest-api-specs/pull/24548/checks?check_run_id=15419664559
> 
> ![image](https://user-images.githubusercontent.com/4429827/260139292-bb41f74c-8af1-4dbc-8f04-e12f00df7c38.png)

I believe this update to the GitHub UI came from a bot calling the GitHub API, which we are trying to avoid.

I believe we can get some basic logging directly into the GitHub UI using the `##[error]` log message.
